Import button

EBro912 opened the pull request:
This PR adds an "Import..." button below the two original buttons in the character selection menu, and clicking this button will allow the user to import VRMs. When a VRM is imported, the file is copied to the mod's `VRM` folder, and a new button is created in the character selection menu, which will automatically open if it is not already open. As of now, the active model **will not** be swapped, and the user will have to click the newly created button to apply the model.

The PR also restores the original functionality of F4, which now simply mimicks the behavior of the import button. Some async methods have also been created to support the Windows file browser in both cases.

EBro912 committed:
create import button

EBro912 committed:
change window title

hrolfurgylfa commented on VRoidHubLoader/Modules/VrmLoaderModule.cs:
```diff
@@ -62,17 +63,37 @@ private void Update()
 
-            // MenuManager is a singleton? sweet.
-            if (!MenuManager.Instance.IsOpen)
+            await ImportVRM();
```
I know it doesn't matter in this case as the Update loop isn't doing anything else, but do we really want to await in Update? Isn't this more of a fire and forget that we don't want to await here.

EBro912 replied:
To my knowledge we kinda have to await here to prevent the file selection window from blocking the thread. I already tried doing this inside of a coroutine and it froze the app. If there's a way to do this without awaiting then I agree.

hrolfurgylfa replied:
Interesting, even if you call ImportVRM without the await statement and keeping ImportVRM as an async function?

I would think that would just fire and forget like JS handles async functions called without await.

EBro912 replied:
I just tried it and the thread does not get blocked if `ImportVRM()` is not awaited, however the current production code does use an `await` before I moved it into a helper function, so we can probably keep it as is if it isn't doing any harm.

hrolfurgylfa replied:
That was code I wrote, but the await was added before merge and I was at work so I didn't see that change until it was already merged, and it wasn't a big enough deal to do another PR to clean it up. I intentionally didn't await there in case more was added to the update loop, as you don't want to block that potential future code, but I probably should have added a comment making that clear.

EBro912 replied:
Ah, thanks for clarifying. Since it seems to work as a fire-and-forget, I can modify it to run that way in a future commit before this is merged.

EBro912 replied:
Changed in [9bcf308](https://github.com/YusufOzmen01/desktopmate-custom-avatar-loader/pull/82/commits/9bcf3089a008107f7b725e3acdc92a9ebd88118c)

hrolfurgylfa commented on VRoidHubLoader/Modules/VrmLoaderModule.cs:
```diff
@@ -62,17 +63,37 @@ private void Update()
+        {
+            string fileName = Path.GetFileName(path);
+            string destination = VrmFolderPath + '\\' + fileName;
+            if (!File.Exists(destination))
```
Isn't it going to be confusing to the user if they import a model but nothing happens? Would it be safe to replace the model, or we could also make a new one with a suffix added to differentiate them like (1), (2) and so on? But that is quite a bit more complex and maybe we just want to replace.

EBro912 replied:
I've been talking with @YusufOzmen01 about this, the loading functionality will most likely be restored here. Although I think the proper way to do this would be to just ignore the load if the model is already selected.

hrolfurgylfa replied:
My worry is with updated models, those would have the same name, but have different content. The user would probably get really confused if they select the new model but then it doesn't have any of the changes they made, and they might think it is a problem on their end.

Also, even if the original loading functionality gets removed, I do really like the idea of also adding it to the VRM folder, that seems really useful.

hrolfurgylfa replied:
If the original loading functionality gets restored, not removed, clearly posted a bit too quickly.

EBro912 replied:
Restored functionality in [0a993f6](https://github.com/YusufOzmen01/desktopmate-custom-avatar-loader/pull/82/commits/0a993f61a4360f3c4ba28ecc3c82f711fe4b98cb)

hrolfurgylfa commented on VRoidHubLoader/Modules/VrmLoaderModule.cs:
```diff
@@ -62,17 +63,37 @@ private void Update()
 
-            // MenuManager is a singleton? sweet.
-            if (!MenuManager.Instance.IsOpen)
+            await ImportVRM();
```
Also, in the case of importing the model with F4, could we also equip the model automatically to keep the old F4 behaviour after it is added to the VRM folder?

EBro912 replied:
Restored functionality in [0a993f6](https://github.com/YusufOzmen01/desktopmate-custom-avatar-loader/pull/82/commits/0a993f61a4360f3c4ba28ecc3c82f711fe4b98cb)

hrolfurgylfa commented on VRoidHubLoader/Modules/VrmLoaderModule.cs:
```diff
@@ -62,17 +63,37 @@ private void Update()
-                #if BEPINEX
-                MonoBehaviourExtensions.StartCoroutine(this, CoAutoOpenModelPage());
-                #endif
+                File.Copy(path, destination);
```
Just another thought, what if this was a symbolic link? That would avoid extra storage space being taken up, especially with bigger 100MB+ models, but it would be less stable and we'd probably need some cleanup logic to delete broken symbolic links. Maybe this is a bigger thing, but might be necessary if we go the suffixing route with duplicate models being added since otherwise the VRM folder might grow a bit too quickly.

EBro912 replied:
I never thought about doing it that way, but in my opinion as of now we don't really have to worry too much about it. We can definitely have this on the backburner but for now I think making a copy or simply moving the file to the VRM folder will suffice.

hrolfurgylfa replied:
Yeah, very true, this is more of a potentially interesting thing to look into in the future if we're bored.

EBro912 committed:
Restore import model loading functionality

Also includes a small hack to avoid an AccessViolationException when attempting to load a model

EBro912 committed:
discard ImportVRM
